clang/test/CheckedC/inferred-bounds/multiple-widened-bounds.c

```diff
@@ -0,0 +1,568 @@
+// Tests for datafow analysis for bounds multiple widening of _Nt_array_ptr's.
+//
+// RUN: %clang_cc1 -fdump-widened-bounds -verify -verify-ignore-unexpected=note -verify-ignore-unexpected=warning %s | FileCheck %s
+
+#include <limits.h>
+#include <stdint.h>
+
+void f1() {
+  _Nt_array_ptr<char> p : count(0) = "a";
+
+  if (*p && *(p + 1)) {}
+
+// CHECK: In function: f1
+// CHECK: [B3]
+// CHECK:   2: *p
+// CHECK: [B2]
+// CHECK:   1: *(p + 1)
+// CHECK: upper_bound(p) = 1
+// CHECK: [B1]
+// CHECK: upper_bound(p) = 2
+}
+
+void f2() {
+  _Nt_array_ptr<char> p : count(0) = "ab";
+
+  if (*p && *(p + 1) && *(p + 2))
+  {}
+
+// CHECK: In function: f2
+// CHECK: [B4]
+// CHECK:   2: *p
+// CHECK: [B3]
+// CHECK:   1: *(p + 1)
+// CHECK: upper_bound(p) = 1
+// CHECK: [B2]
+// CHECK:   1: *(p + 2)
+// CHECK: upper_bound(p) = 2
+// CHECK: [B1]
+// CHECK: upper_bound(p) = 3
+}
+
+void f3() {
+  _Nt_array_ptr<char> p : count(0) = "a";
+  int a;
+
+  if (*p && *(p + 1)) {
+    p = "a";
+    if (a) {}
+  }
+
+// CHECK: In function: f3
+// CHECK: [B4]
+// CHECK:   3: *p
+// CHECK: [B3]
+// CHECK:   1: *(p + 1)
+// CHECK: upper_bound(p) = 1
+// CHECK: [B2]
+// CHECK:   1: p = "a"
+// CHECK: upper_bound(p) = 2
+// CHECK: [B1]
+// CHECK-NOT: upper_bound(p) = 1
```

Comment:
We can replace these two `CHECK-NOT`s with a single `CHECK-NOT: upper_bound(p)`.